src/network/net_addons.cc

```diff
@@ -38,309 +53,761 @@
 
 namespace AddOns {
 
-// silence warnings triggered by curl.h
-CLANG_DIAG_OFF("-Wdisabled-macro-expansion")
+/*
+ * The networking-related code in this file is inspired by
+ * https://www.thecrazyprogrammer.com/2017/06/socket-programming.html
+ *
+ * The communication protocol is documented in the server
+ * repo (widelands/wl_addons_server) in `Server.java`.
+ */
+
+namespace {
 
-// all CURL-related code is inspired by
-// https://stackoverflow.com/questions/1636333/download-file-using-libcurl-in-c-c
+inline int portable_write(const int socket, const char* buffer, const size_t length) {
+#ifdef _WIN32
+	return send(socket, buffer, length, 0);
+#else
+	return write(socket, buffer, length);
+#endif
+}
+inline int portable_read(const int socket, char* buffer, const size_t length) {
+#ifdef _WIN32
+	return recv(socket, buffer, length, 0);
+#else
+	return read(socket, buffer, length);
+#endif
+}
 
-void NetAddons::init() {
-	if (curl_) {
+inline void check_string_validity(const std::string& str) {
+	if (str.find(' ') != std::string::npos) {
+		throw WLWarning("", "String '%s' may not contain whitespaces", str.c_str());
+	}
+	if (str.find('\n') != std::string::npos) {
+		throw WLWarning("", "String '%s' may not contain newlines", str.c_str());
+	}
+}
+}
+
+constexpr unsigned kCurrentProtocolVersion = 4;
+
+void NetAddons::init(std::string username, std::string password) {
+	if (initialized_) {
 		// already initialized
 		return;
 	}
-	curl_ = curl_easy_init();
+	if (network_active_) {
+		throw WLWarning("", "Network is already active during init");
+	}
 
-#ifdef __MINGW32__
-	// mingw with static curl causes some trouble
-	curl_easy_setopt(curl_, CURLOPT_SSL_VERIFYPEER, 0);
+#ifdef SIGPIPE
+	signal(SIGPIPE, SIG_IGN);  // NOLINT
 #endif
 
-	if (!curl_) {
-		throw wexception("Unable to initialize CURL");
+	if (password.empty()) {
+		username = "";
 	}
-}
+	if (username.empty()) {
+		username = last_username_;
+	} else {
+		last_username_ = "";
+	}
+	if (password.empty()) {
+		password = last_password_;
+	} else {
+		last_password_ = "";
+	}
+	check_string_validity(username);
 
-NetAddons::~NetAddons() {
-	if (curl_) {
-		curl_easy_cleanup(curl_);
-		curl_ = nullptr;
+	if ((client_socket_ = socket(AF_INET, SOCK_STREAM, 0)) < 0) {
+		throw WLWarning("", "Unable to create socket");
+	}
+	sockaddr_in server;
+	server.sin_family = AF_INET;
+	server.sin_port = htons(get_config_int("addon_server_port", 7399));
+	server.sin_addr.s_addr =
+	   inet_addr(get_config_string("addon_server_ip", "127.0.0.1" /* NOCOM */).c_str());
```

Comment:
Do you plan to paste the static ip of the server here later on, or should this also support name resolution?